source/common/network/connection_impl.cc

```diff
@@ -34,6 +34,16 @@ Address::InstanceConstSharedPtr getNullLocalAddress(const Address::Instance& add
 }
 } // namespace
 
+int ConnectionImplUtility::createSocket(Address::InstanceConstSharedPtr address,
+                                        Optional<Address::InstanceConstSharedPtr> source_address) {
+  int fd = address->socket(Address::SocketType::Stream);
+  if (fd >= 0 && source_address.valid()) {
+    int rc = source_address.value()->bind(fd);
+    ASSERT(rc >= 0);
```

Comment:
It should definitely be stronger than ASSERT(). We could end up routing traffic incorrectly if this call fails to set the source address so a silent failure is not good.